core/src/main/scala/com/mesosphere/usi/core/Scheduler.scala

```diff
@@ -57,6 +59,137 @@ object Scheduler {
   def fromSnapshot(specsSnapshot: SpecsSnapshot, client: MesosClient): Flow[SpecUpdated, StateOutput, NotUsed] =
     Flow[SpecUpdated].prefixAndTail(0).map { case (_, rest) => specsSnapshot -> rest }.via(fromClient(client))
 
+  def asFlow(specsSnapshot: SpecsSnapshot, client: MesosClient)(
+      implicit materializer: Materializer): Future[(StateSnapshot, Flow[SpecUpdated, StateEvent, NotUsed])] = {
+
+    implicit val ec = scala.concurrent.ExecutionContext.Implicits.global //only for ultra-fast non-blocking onComplete
+
+    val (snap, source, sink) = asSourceAndSink(specsSnapshot, client)
+
+    snap.map { snapshot =>
+      (snapshot, Flow.fromSinkAndSourceCoupled(sink, source))
+    }
+
+  }
+
+  /**
+    * Represents the scheduler as a Sink and Source.
+    *
+    * This method will materialize the scheduler first, then Sink and Source can be materialized independently, but only once.
+    *
+    * @param specsSnapshot Snapshot of the current specs
+    * @return Snapshot of the current state, as well as Source which produces StateEvents and Sink which accepts SpecEvents
+    */
+  def asSourceAndSink(specsSnapshot: SpecsSnapshot, client: MesosClient)(
+      implicit mat: Materializer): (Future[StateSnapshot], Source[StateEvent, NotUsed], Sink[SpecUpdated, NotUsed]) = {
+    val flow = fromClient(client)
+    asSourceAndSink(specsSnapshot, flow)(mat)
+  }
+
+  def asSourceAndSink(specsSnapshot: SpecsSnapshot, schedulerFlow: Flow[SpecInput, StateOutput, NotUsed])(
+      implicit mat: Materializer): (Future[StateSnapshot], Source[StateEvent, NotUsed], Sink[SpecUpdated, NotUsed]) = {
+
+    implicit val ec = scala.concurrent.ExecutionContext.Implicits.global //only for ultra-fast non-blocking onComplete
+
+    val (stateQueue, stateSource) = Source.queue[StateEvent](1, OverflowStrategy.backpressure).preMaterialize()
+
+    val (specQueue, specSink) = Sink.queue[SpecUpdated]().preMaterialize()
+
+    val stateSnapshotPromise = Promise[StateSnapshot]()
+
+    val killSwitch = KillSwitches.shared("SchedulerAdapter.asSourceAndSink")
```

Comment:
This whole `killSwitch` business is lamentable. Ideally, we should not need it at all.